### ISSUE TITLE
should ignore BOM in .gitignore

### ISSUE THREAD

**Issue by skelter** [libgit2/libgit2#5984]
Manually migrating from https://github.com/libgit2/libgit2sharp/issues/1903

### Reproduction steps
1. Try this with an existing repo. Put a BOM at the header of a .gitignore file, and a directory to ignore such as .vs/ on the same (top, first) line of the .gitignore file.
2. confirm command-line native git still ignores the .vs directory.
3. Write a quick program that uses libgit2sharp to look at the above repo and get the status.
4. Get the status and note the status is marked dirty, and the .vs dir is listed as untracked, not ignored.

### Expected behavior
The status reported by libgit2sharp should report status is clean. It should ignore the .vs dir.

### Actual behavior
The .vs dir is listed as untracked, not ignored.

The status is dirty.

### Version of libgit2 (release number or SHA1)
the version in LibGit2Sharp.NativeBinaries v 2.0.306. that was referenced by LibGit2Sharp 0.26.2

### Operating system(s) tested
Windows 10

**Comment by bording** [comment]
> the version in LibGit2Sharp.NativeBinaries v 2.0.314. that was referenced by LibGit2Sharp 0.26.2

FYI, LibGit2Sharp 0.26.2 references LibGit2Sharp.NativeBinaries 2.0.306, not 2.0.314. The libgit2 hash in 2.0.306 is 106a5f27586504ea371528191f0ea3aac2ad432b.

**Comment by skelter** [comment]
bording, you are right.  I was reading the NuGet UI incorrectly.

**Comment by ethomson** [comment]
Thanks @skelter for reporting and sorry for the trouble.  Adding some labels in case a new contributors wants to pick this up, we have functions that will skip a bom so it should be relatively straightforward fix!  (Bonus points for auditing `.gitattributes` and `.git/config` to see if git has changed behavior wholesale on all of these sorts of config files, or if it's piecemeal.  Bonus bonus points for making a well reasoned argument that we should break behavior with git and skip the bom on all of these.  😁)

**Comment by skelter** [comment]
ethomson, and all, thank you for providing a very useful library and all the effort y'all have put into it.

**Comment by suehshtri** [comment]
Further discussion in git-scm context at https://github.com/git/git-scm.com/issues/663

**Comment by stforek** [comment]
This looks to be resolved in this comtmit: https://github.com/libgit2/libgit2/commit/133bcebad9af73395f1c1e09f85e7e980bd20744
At least I can't reproduce.

**Comment by suehshtri** [comment]
> This looks to be resolved in this comtmit: [133bceb] ... At least I can't reproduce.
@stforek , I can be available to help with repro.

**Comment by stforek** [comment]
@suehshtri can you provide .gitignore which libgit2 fails to parse?

**Comment by stforek** [comment]
> > the version in LibGit2Sharp.NativeBinaries v 2.0.314. that was referenced by LibGit2Sharp 0.26.2
> 
> FYI, LibGit2Sharp 0.26.2 references LibGit2Sharp.NativeBinaries 2.0.306, not 2.0.314. The libgit2 hash in 2.0.306 is [106a5f2](https://github.com/libgit2/libgit2/commit/106a5f27586504ea371528191f0ea3aac2ad432b).

I even built libgit2 from this commit and still can't reproduce with this [gitignore.txt](https://github.com/libgit2/libgit2/files/7641069/gitignore.txt) (please rename it to .gitignore).


```
#include <iostream>
#include <git2.h>

int status_cb(const char *path, unsigned int status_flags, void *payload) {
    std::cout << path << '\n';
    return 0;
}

int main(int argc, char **argv) {
    git_libgit2_init();

    git_repository* repo;
    int err = git_repository_open(&repo, "/home/pc/workspace/libgit2_test");
    if (err) {
        std::cout << "git_repository_open error: " << err << "\n";
    }
    git_status_options opts = GIT_STATUS_OPTIONS_INIT;
    opts.show  = GIT_STATUS_SHOW_INDEX_AND_WORKDIR;
    opts.flags = GIT_STATUS_OPT_INCLUDE_UNTRACKED |
			GIT_STATUS_OPT_RECURSE_UNTRACKED_DIRS;
    git_status_foreach_ext(repo, &opts, status_cb, nullptr);

    git_repository_free(repo);
    git_libgit2_shutdown();
    return 0;
}

```

**Comment by suehshtri** [comment]
I will see if I can get you the actual file that showed the problem for us.
In the mean time, at this revision we had the problem:

```````
00000000: efbb bf2f 2e76 730d 0a2f 4261 7365 6c69  .../.vs../Baseli
00000010: 6e65 2f41 6374 7561 6c0d 0a              ne/Actual..
`````````

and at this revision, we dodged the problem:
``````
00000000: efbb bf23 2069 676e 6f72 6520 5669 7375  ...# ignore Visu
00000010: 616c 5374 7564 696f 2066 696c 6573 0d0a  alStudio files..
00000020: 2f2e 7673 0d0a 2f42 6173 656c 696e 652f  /.vs../Baseline/
00000030: 4163 7475 616c 0d0a                      Actual..
````````

**Comment by ethomson** [comment]
🤔  That's interesting - since both files have the BOM, it suggests that it might be a more subtle problem.

**Comment by suehshtri** [comment]
I attach a zip with two files: 
[dotgitignore-bom-examplefiles.zip](https://github.com/libgit2/libgit2/files/7642585/dotgitignore-bom-examplefiles.zip)

**Comment by suehshtri** [comment]
Ed, I agree.  I wonder ... does it have more to do with it being the first line, or an immediate directory with a dot?  We dodge it by making the first line a comment, or at least that is our mental model of it (which could be wrong, but resulted in useful behavior).

**Comment by ethomson** [comment]
Right!  I'm stumped.  Those two certainly _shouldn't_ be different, but you know that already.  😁 

We'll have to step through in a debugger.  😓

**Comment by stforek** [comment]
> I attach a zip with two files: [dotgitignore-bom-examplefiles.zip](https://github.com/libgit2/libgit2/files/7642585/dotgitignore-bom-examplefiles.zip)

I created a repo with .gitignore from this zip (dotgitignore-causesproblem). This is my dir structure:

```
.gitignore
.vs/test.txt
```

git_status_foreach_ext() doesn't list this file. (opts.flags = GIT_STATUS_OPT_INCLUDE_UNTRACKED |GIT_STATUS_OPT_RECURSE_UNTRACKED_DIRS;)
git_status_file(&status, repo, ".vs/test.txt") returns status with GIT_STATUS_IGNORED flag.
git_status_list_new() doesn't list this file.
git_status_should_ignore(&ignored, repo, ".vs/test.txt") returns ignored = 1.

I'm trying this on linux. Maybe it can be reproduced only under Windows?

**Comment by suehshtri** [comment]
stforek, I have not tried it on Linux.  We see it on Windows.

**Comment by csware** [comment]
I suppose this is fixed (cf. 73a157e0df2b562889ec62326c66bdaf0b0dfd81)

**Comment by suehshtri** [comment]
> I suppose this is fixed (cf. [73a157e](https://github.com/libgit2/libgit2/commit/73a157e0df2b562889ec62326c66bdaf0b0dfd81))

I will test this today.
